sno/merge.py

```diff
@@ -207,24 +135,76 @@ def move_repo_to_merging_state(repo, merge_index, merge_context, merge_message):
     write_repo_file(repo, MERGE_MSG, merge_message)
 
 
-def abort_merging_state(repo):
+def abort_merging_state(ctx):
     """
     Put things back how they were before the merge began.
     Tries to be robust against failure, in case the user has messed up the repo's state.
     """
+    repo = ctx.obj.repo
     is_ongoing_merge = repo_file_exists(repo, MERGE_HEAD)
     # If we are in a merge, we now need to delete all the MERGE_* files.
     # If we are not in a merge, we should clean them up anyway.
-    remove_repo_file(repo, MERGE_HEAD)
-    remove_repo_file(repo, MERGE_MSG)
-    remove_repo_file(repo, MERGE_INDEX)
-    remove_repo_file(repo, MERGE_LABELS)
+    remove_all_merge_repo_files(repo)
 
     if not is_ongoing_merge:
         raise InvalidOperation('Repository is not in "merging" state.')
+    else:
+        ctx.exit(0)
+
 
-    # TODO - maybe restore HEAD to ORIG_HEAD.
-    # Not sure if it matters - we don't modify HEAD when we move into merging state.
+def complete_merging_state(ctx):
+    """
+    Completes a merge that had conflicts - commits the result of the merge, and
+    moves the repo from merging state back into the normal state, with the branch
+    HEAD now at the merge commit. Only works if all conflicts have been resolved.
+    """
+    repo = ctx.obj.repo
+    if not is_ongoing_merge(repo):
+        raise InvalidOperation('Repository is not in "merging" state.')
+    merge_index = MergeIndex.read_from_repo(repo)
+    if merge_index.unresolved_conflicts:
+        raise InvalidOperation(
+            "Merge cannot be completed until conflicts are resolved."
+        )
+
+    merge_context = MergeContext.read_from_repo(repo)
+    commit_ids = merge_context.versions.map(lambda v: v.repo_structure.id)
+    merge_message = read_repo_file(repo, MERGE_MSG)
```

Comment:
User doesn't get prompted to edit this like a normal commit?

Reply:
... I guess they should? I copied the existing behaviour from sno merge (no conflicts). I'll fix both in a follow up PR